### ISSUE TITLE
Help: download_attribute takes a long time to return?

### ISSUE THREAD

**Issue by yxiang92128** [Azure/azure-storage-cpp#320]
When we download a block blob, we need to know it's last modified time and also the size of the blob so I perform a 
block_blob.download_attributes() 
But very often, we see that call being blocked and waited for several minutes to come back and thus severely impact the performance of overall operation including the subsequent download_range_to_stream call.
I wonder if there is an alternative to retrieve the last modified time and the size of the blob much more quickly and if there is a workaround?
Here is a stack trace showing where it blocks and waits:
Stack Trace: Thread 8131 (0x1fc3), Process 7512 (0x1d58), actmain
00 0x00007ffff13ff0bf Frame<00007ffff13ff0bf> Return<0000000000000000> libpthread.so.0!pthread_cond_wait@@GLIBC_2.3.2 + 0xbf
01 0x00007fffeab9b87c Frame<00007fffeab9b87c> Return<0000000000000000> libstdc++.so.6!_ZNSt18condition_variable4waitERSt11unique_lockISt5mutexE + 0xc
02 0x00007fff9534e4eb Frame<00007fff9534e4eb> Return<0000000000000000> _ZNSt18condition_variable4waitIZN4pplx7details10event_impl4waitEjEUlvE_EEvRSt11unique_lockISt5mutexET_ + 0x29 condition_variable:93
03 0x00007fff9534454e Frame<00007fff9534454e> Return<0000000000000000> pplx::details + 0x46 pplxlinux.h:98
04 0x00007fff95344615 Frame<00007fff95344615> Return<0000000000000000> pplx::details + 0x1d pplxlinux.h:113
05 0x00007fff95345a3e Frame<00007fff95345a3e> Return<0000000000000000> pplx::details + 0x18 pplx.h:161
06 0x00007fff95345a24 Frame<00007fff95345a24> Return<0000000000000000> pplx::details + 0x18 pplx.h:158
07 0x00007fff9534658d Frame<00007fff9534658d> Return<0000000000000000> pplx::details + 0x45 pplxtasks.h:1777
08 0x00007fff9534f954 Frame<00007fff9534f954> Return<0000000000000000> _ZNK4pplx4taskIhE4waitEv + 0x6c pplxtasks.h:3601
09 0x00007fff95347990 Frame<00007fff95347990> Return<0000000000000000> _ZNK4pplx4taskIvE4waitEv + 0x18 pplxtasks.h:4594
10 0x00007fff9534ac2e Frame<00007fff9534ac2e> Return<0000000000000000> azure::storage + 0x2c blob.h:4502
11 0x00007fff953403ff Frame<00007fff953403ff> Return<0000000000000000> AZ_blob::download_blob_operation + 0x3b3 AZ_blob.cpp:443

A snippet of code where I call download_attributes:
   azure::storage::cloud_block_blob block_blob = container.get_block_blob_reference(obj_name);
    azure::storage::blob_request_options reqOptions;
    reqOptions.set_parallelism_factor(1);
    reqOptions.set_maximum_execution_time(std::chrono::milliseconds(iface->max_wait_time));
    reqOptions.set_server_timeout(std::chrono::seconds(iface->connect_ms/1000));

    // increase read buffer to 16M in order to give a boost to performance
    const size_t new_stream_read_size = 16 * 1024 * 1024;
    reqOptions.set_stream_read_size_in_bytes(new_stream_read_size);

    //need to obtain last modified time
    block_blob.download_attributes();
block_blob.properties().last_modified().to_string().c_str());
    long unsigned int input = block_blob.properties().last_modified().to_interval();
    long unsigned int linuxtime_milisecs = input/10000 - epoch_offset;// diff between windows and unix epochs (seconds)

Thanks!

**Comment by JinmingHu-MSFT** [comment]
Hi @yxiang92128 , blocking at `download_attributes` seems unusual, because unlike download/upload blob operations, this is a very simple function.

Can you provide more information so that I can investigate further?

1. Does this only block at `download_attributes` function, not other functions, not even once?

2. Which version are you using? Because I see `blob.h:4502` in your stacktrace, and I don't know which function it maps to. Because `blob.h` changes a lot.

3. Since you mentioned `very often`, can you stably reproduce it with short code which is easy to understand and works in common environment?

**Comment by yxiang92128** [comment]
Hi @JinmingHu-MSFT 
Let me rephrase my question:
Is there a way to download the metadata such as size of the blob and last modified timestamp along with the bytes stream for the actual blob content with one call?  i.e. one REST roundtrip to the backend?  I measured it and the download_attributes call for the blob accounts 20% of the time spent along with a separate download_stream_to_buffer call.
Any workaround?  We are looking at any performance boost we can gain by tweaking the current SDK operations.  

Thanks,

Yang

**Comment by JinmingHu-MSFT** [comment]
Hi @yxiang92128 , yes, as long as you've downloaded a part of a file, the properties and metadata are also downloaded and filled in the `cloud_blob` object. You can access them with `properties()` and `metadata()`.

**Comment by yxiang92128** [comment]
@JinmingHu-MSFT  that helps and worked. Thanks!!
